_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,429 @@
+}
+
+contract TestBal is DSTest, SafeMath {
+    function prove_transfer(address dst, uint amt) public {
+        // BAL: https://etherscan.io/address/0xba100000625a3754423978a60c9317c58a424e3D#code
+        ERC20 token = ERC20(0xba100000625a3754423978a60c9317c58a424e3D);
+
+        // ignore cases where we don't have engough tokens
+        if (amt > bal.balanceOf(address(this))) return;
+
+        uint preBalThis = token.balanceOf(address(this));
+        uint preBalDst  = token.balanceOf(dst);
+
+        token.transfer(dst, amt);
+
+        // no change for self-transfer
+        uint delta = dst == address(this) ? 0 : amt;
+
+        // balance of `this` has been reduced by `delta`
+        assertEq(sub(preBalThis, delta), bal.balanceOf(address(this)));
+
+        // balance of `dst` has been increased by `delta`
+        assertEq(add(preBalDst, delta), bal.balanceOf(dst);
+    }
+}
+```
+
+Lets run this test as the [balancer DAO](https://etherscan.io/token/0xba100000625a3754423978a60c9317c58a424e3d?a=0xb618f903ad1d00d6f7b92f5b0954dcdc056fc533) to make sure that we have plenty of `BAL`:
+
+```
+$ DAPP_TEST_ADDRESS=0xb618f903ad1d00d6f7b92f5b0954dcdc056fc533 dapp test --rpc-url <URL>
+[FAIL] prove_transfer(address,uint256)
+
+Failure: prove_transfer(address,uint256)
+
+  Counter Example:
+
+    result:   Revert("ERC20: transfer to the zero address")
+    calldata: prove_transfer(0x0000000000000000000000000000000000000000, 0)
+```
+
+We have uncovered another edge case! The balancer token disallows transfers to the zero address.
+
+It is also worth noting that symbolic execution against rpc state can be significantly more
+performant than fuzzing via rpc.
+
+## Interactive Exploration
+
+`hevm` also includes a visual debugger, and we can use this to interactively explore the execution
+tree. You can enter the debugger by running `dapp debug` from the root of your `dapp project`. You
+will see a list of test methods, and once you select one you will be dropped into an interactive
+debugging session.
+
+You can press `h` to bring up a help view, `n` to step forwards, and `p` to step back. If you press
+`e` in symbolic test you will jump to the next branching point, once there you can press `0` to choose
+the branch which does not jump, and `1` to choose the branch that does.
+
+Note that the interactive debugger will also function when executing against mainnet state.
+
+A small demonstation video can be found below:
+
+
+
+
+## Limitations, Assumptions & Future Work
+
+#### Non Linearity (`safeMul`)
+
+The symbolic execution engine in `hevm` is backed by an [SMT
+solver](https://en.wikipedia.org/wiki/Satisfiability_Modulo_Theories) (currently either `z3` or
+`cvc4` are supported). Expressions involving non-linear arithmetic (multiplication or divison by
+symbolic variables) are extremely challenging for SMT solvers, and it will often not be possible to
+symbolically execute tests involving lots of non linear arithmetic.
```

Comment:
Non-linear is written `non-linear` above and `non linear` here. In a quick search, I mainly find `non-linear` and `nonlinear` as referred by tools and papers.